Functional Component Refactor

prkirby opened the pull request:
Refactor PtsCanvas into a functional component. This also gets rid of the need for QuickStartCanvas, as they serve the same function. 

Major Changes:

- PtsCanvas is now a pure functional component
- Uses hooks
- All IPlayer handlers are passed as props
- Uses useRef hook to maintain reference to Space and Form
- Optional Tempo prop to allow users to create a Tempos in their parent component and have it start with this. 

References #15

williamngan commented:
Thanks @prkirby -- I'll review this soon.

One quick question: would it be possible to keep the old `class` approach here for legacy support?

tibotiber commented:
Hi @prkirby, @williamngan,

I'm back in action, I've checked the code and tested the PR branch from within my app.

On the code side, it's pretty much the way I expected it to be, nothing to comment (at least on the JS side, not familiar with TS). I like how the 2 legacy components are refactored as a single functional component.

On the testing side, 2 things stood out:
1. The refactor from an `extends` of PtsCanvas class component to a parent component of the new functional component is not trivial (a few hours for my rather complex component) so I support @williamngan's suggestion to keep the current class component for legacy support (at least for a while). 
2. One use-case is not currently supported. I need access to the canvas element in my parent component, I could do that easily in the past as class components can take their ref from the parent using `React.forwardRef`, this is not possible now. So I suggest that we add a `forwardRef` to the PtsCanvas component and pass the ref to the canvas element. I am not sure of the behaviour if you use a `forwardRef` and the parent does not pass a `ref`, you might need to test for that and create it if it's not passed.

I hope this helps, let me know if some things are not clear. Also please ping me so I can test any new update ;). And I almost forgot, thanks for the PR — great job. 🎉

tibotiber commented:
Forgot a nitpick. Should `tempo` be added to `defaultProps`?

tibotiber commented:
Thanks @williamngan, yes the break was way overdue and great.

For `forwardRef`, it's not a hook, it needs to wrap the component, similarly to `React.memo`. It's a bit of a special one and can only be done in the lib itself from what I've gathered.

prkirby commented:
@tibotiber @williamngan thanks for the feedback! Yes I will put the legacy class based component back in.  

@tibotiber so, in your parent component, will you be using the useRef() hook? And then you want to be able to pass that ref down to the canvas element in the Pts functional component, so you can access it from the parent? Cause if thats the case I think I can just use a simple OR clause to either assign the ref to a new useRef() hook in the functional component, or use the ref passed as a prop. Something like: 

```javascript
const canvRef = props.canvRef || useRef(null)
```

tibotiber commented:
@prkirby yes that's exactly how I intend to use it and how I imagined you would do the implementation :)

prkirby commented:
@tibotiber Can you take a look now and see if the canvRef prop works for you? I also added the option to pass a spaceRef and formRef if there was ever a desire to do so. 

@williamngan I added back in the class based component, and updated the README. Let me know how things look.

tibotiber commented:
Hi @prkirby, thanks for the further work on this. A few things:

1. Exposing refs to DOM nodes via props doesn't work for functional components. I think it has something to do with the way refs are managed internally by React. You can read more about it [here](https://reactjs.org/docs/refs-and-the-dom.html#exposing-dom-refs-to-parent-components). TLDR: you need to use `forwardRef` to wrap the component and use `ref` instead of canvRef. Note that this is only supported from react 16.3 onwards.
2. I'm not sure how the space and form refs would work since they are not DOM node refs, but in any case a parent component could rely on `onStart` to get the references, maybe that's good enough?
3. The separation of the legacy class component and the new functional component is good. What would you and @williamngan think of renaming the legacy component, calling `PtsCanvas` the functional component (since more in line with modern React), and bumping the version to highlight a breaking change?

Hope this helps, thanks for the work here :)

williamngan commented:
Thank you so much, @prkirby and @tibotiber . I would support renaming the legacy component (maybe calling it PtsCanvasElement? Open to suggestions) and yes I'll bump the version.

williamngan commented:
Hi @prkirby @tibotiber , I tried this PR and ran into two (probabbly config) issues.

1. I created a new project via create-react-app, `npm link react-pts-canvas`, and then try this test code:
```jsx
function App() {
  return (
    <div className="App">
      <PtsCanvasFC name="hello" onAnimate={
        (space,form) => {
          form.point( space.pointer, 10 );
        }
      } />
    </div>
  );
}
```
I got an error about "Invalid hook call". Some forums suggested this might be related to `npm link`. Did you encounter the same and any suggestions for a fix?

2. My VSCode doesn't seem to recognize `index.d.ts` and highlight typing errors in `index.ts`. Any tips?

tibotiber commented:
@williamngan I didn't encounter this, but I'm using `yarn` with `yarn workspaces` for local dependencies, I have never even heard of `npm link` 😬. Also not a TS user, so I'm not familiar with point 2 either, sorry.

williamngan commented:
Hahaha, nice one @tibotiber . I should start using yarn workspaces but I'll keep VS Code. :)

Are we good with this PR? If so, I'll merge it and then patch if necessary.

Thanks guys!

tibotiber commented:
I'll definitely back you up on VS Code, but not the TS part (yet!).

For the PR, it's not good to go yet, still pending changes based on my comment.

tibotiber commented:
Hi @prkirby, just come across `useImperativeHandle`, see [docs](https://reactjs.org/docs/hooks-reference.html#useimperativehandle) and realised that if you wanted to expose all refs (canvas, space and form), this could be the most React-ish way to do it. Your ref would now be an object that includes the 3 refs. No a need from me, just food for thoughts :)

prkirby commented:
Awesome! I will take a look at it, it sounds perfect. Sorry for the delay,
I’ve been in the process of moving so I’m a bit behind. Will check it out
this week.

On Mon, Mar 22, 2021 at 12:30 AM Thibaut Tiberghien <
***@***.***> wrote:

> Hi @prkirby <https://github.com/prkirby>, just come across
> useImperativeHandle, see docs
> <https://reactjs.org/docs/hooks-reference.html#useimperativehandle> and
> realised that if you wanted to expose all refs (canvas, space and form),
> this could be the most React-ish way to do it. Your ref would now be an
> object that includes the 3 refs. No a need from me, just food for thoughts
> :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/williamngan/react-pts-canvas/pull/32#issuecomment-803772590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGGR4IKVCQWVKHPXKFWJVTTE3IXNANCNFSM4XQD3DPQ>
> .
>
-- 
Regards,

*Paul Kirby*
*Web Developer / Audio Engineer*
*Chicago, IL*

tibotiber commented:
Take your time, your contribution is already appreciated so it's on your terms ;)

williamngan commented:
Good luck with the move @prkirby . No problem at all and thank you for making this PR.

williamngan commented:
Happy New Year! Hey @prkirby - just want to check if you're still planning to update this PR. 

No a big rush, but if you're not planning to update it, is it ok if I merge it as is and see if I can fix the remaining issues?

I got some questions for React support from users so I'd like to get it out :)

prkirby commented:
Hey @williamngan , sorry about that totally slipped my mind. Let me take a crack at it this weekend, I've got some spare time.

williamngan commented:
Thank you so much @prkirby !

prkirby commented:
Hey @williamngan I wasn't able to get to it this past weekend. If you want to merge it and see if you can handle the forwardRef issue feel free, otherwise I may be able to poke at it later this week / weekend.

prkirby commented:
@williamngan spent the day finalizing the PR. Took the time to modernize the TypeScript environment, and had to do a good deal of updating / maintenance to get everything running with rollup and react-scripts inside example

I know its a lot to review haha, but I figured this puts you in a good place to update / add features and not have to worry about maintaining type files manually, as well as standardizing how the TS files will be linted in IDE with the .tsconfig set up.

@tibotiber I ended up using `forwardRef` just for the canvas ref, as you were right in assuming that its only DOM refs that can't be passed down as props. Any other mutable type is fine, as far as I can tell. I added an example file under example/App-ParentRef if you want to see it in action.

Any questions or requests lmk

williamngan commented:
Wow, thank you so much @prkirby ! I'll review it soon in the next 1-2 days.

If @tibotiber has time for a quick review, that would be amazing too.

Thanks everyone ❤️

tibotiber commented on src/index.tsx:
```diff
@@ -0,0 +1,172 @@
+  )
+}
+
+export const PtsCanvas = forwardRef<HTMLCanvasElement, PtsCanvasProps>(
```
aren't HTMLCanvasElement and PtsCanvasProps in the opposite order?

prkirby replied:
Haha I thought the same thing, but no they are reversed for some reason, https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/forward_and_create_ref/

Ref type first, then props type.

tibotiber commented on src/index.tsx:
```diff
@@ -0,0 +1,172 @@
+  onResize?: HandleResizeFn
+  onAction?: HandleActionFn
+  tempo?: Tempo
+  canvRef?: React.MutableRefObject<HTMLCanvasElement | undefined>
```
Should this be removed?

tibotiber reviewed (commented):
Hey, I've started testing this. It doesn't work yet in my app but I'm not sure why (could be on my end). I'm sharing a few comments on stuff I found curious in the code. I haven't gone through properly though, just wanted to highlight that so I don't forget.

tibotiber commented:
Hi @prkirby, I'm unfortunately not able to get this to work 100% in my app. It generally works for rendering and interaction, however the forwardRef part breaks some functionality. I've checked the implementation and it looks good to me. I've also checked that the ref I access in the parent component is looking ok and it does. Below is a description of my bug:

My parent component has a keydown handler that clicks the canvas whenever the return key is pressed. Before this PR, pressing return will act exactly like clicking at the current mouse position. After this PR, the click is broken and the mouse jumps out of the canvas (maybe it goes to 0,0?). 

Do you have any idea what may be happening? or maybe @williamngan? I would love to create a minimal repro case but i'm really swamped here, any chance one of you might take a chance at it?

williamngan commented:
Hi @prkirby and @tibotiber -- first of all, thanks for everything! I reviewed the code as best as I can. It looks good but I felt like I'm not competent in the new React patterns to dive deeper. 

I'll try to create a mini-repo this week for us to collaborate on. Sorry for the delay here -- we just moved and are still unpacking boxes :)

williamngan commented on example/src/App-ParentRef.tsx:
```diff
@@ -0,0 +1,103 @@
+  // Doesn't re-render based on the radius, it only affects the handleAnimate
+  const radiusRef = useRef(50)
+  const canvRef = createRef<HTMLCanvasElement>()
+  const spaceRef = useRef<CanvasSpace>()
```
The `spaceRef` and `formRef` really confuses me since I'm not yet familiar with the React patterns. Maybe we can add more comments here to help user understand how this works?

tibotiber replied:
I'm wondering if these props are really needed. If the parent component needs to access the space or the form, they can get it from `onStart`, isn't it?

prkirby replied:
@tibotiber that is true, I suppose the difference is whether or not they want access to the actual ref being used within the component? I guess there aren't to many use cases for that, so happy to remove

prkirby replied:
Again, seems like a rare use case and just makes things more confusing. Removed for now.

williamngan commented on example/src/App.tsx:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  const handleAnimate: HandleAnimateFn = (space, form, time) => {
+    if (!form || !space || !gridRef.current || time === undefined) return
```
Do we always need to check for undefined here? If I understand correctly, this is due to the optional type declaration. I wonder if there's a better way to make defaults so that we don't need do this check. Thoughts?

prkirby replied:
@tibotiber @williamngan Thats really just a design choice, up to you! By having them be optional and letting users decide how to handle undefined situations lets them write some other side effect if they want to when space or form, etc, are undefined. 

Otherwise, yeah @tibotiber we can just not run handleAnimate if those are undefined, and then users don't have to check. Let me know what way you want to go with that!

williamngan commented on example/src/App.tsx:
```diff
@@ -0,0 +1,106 @@
+const ExampleComponent2: React.FC = () => {
+  // Use a useRef instead of useState hook here as the canvas
+  // Doesn't re-render based on the radius, it only affects the handleAnimate
+  const radiusRef = useRef(50)
```
This ref approach for mutable variables is probably correct, but it feels heavy... is there an alternative approach where we can use `let radius = 50`? If there is, it would be good to have an example here too.

williamngan reviewed (commented):
I added a couple thoughts here, mostly thinking about how we can make it easier for users who may not be an expert in React patterns. (eg, myself 😄)

prkirby commented:
Added a minimal repro for the canvRef issue you were seeing, Its in App-ParentRef. I'm seeing the click even move mouse position to 0,0. I've never forced a click event before so I'm not 100% sure what expected behavior or how to debug. will dig in

prkirby commented:
@tibotiber I'd have to see what your original code is, I'm not seeing anything that would make HTMLElement.click() method pay attention to where your mouse position is, it looks like it only forces a generic click event on that element? https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/click

tibotiber commented:
Unfortunately my code is closed source, but if your minimal repro already has the behaviour of jumping to 0,0 it's a good sign. Would you have the time to convert your minimal repro to the legacy component and see if there is any difference?

williamngan commented:
Hey @prkirby and @tibotiber - I took a look at this bug. It's interesting. 

First, I tested it without React and can reproduce the (0,0) issue in vanilla js. So looks like this is the default behavior of canvasElement.click() and not a bug from our side. My rough test code:

```jsx
<html>
<head></head>
<body>
  <canvas id="pts" width="500" height="500" style="box-shadow: 0 0 10px;"></canvas>
  <script>
    const canv = document.querySelector("#pts");
    function handleClick(e) {
      console.log( e.clientX, e.clientY );
      canv.getContext('2d').strokeRect(e.clientX, e.clientY, 10, 10);
    }
    canv.addEventListener('click', handleClick);
    window.addEventListener('keyup', function(e) {
      if (e.key === 'Enter') {
        canv.click();
      }
    })
  </script>
</body>
</html>
```

Then for the legacy class component, I also encountered this issue (Not sure how @tibotiber get his to work). But I can hack it by adding the listener inside the `start` function, so that I can pass `space.pointer` into it. That's probably not the ideal React way, but it works :)

```tsx
class ExampleComponent extends PtsCanvasLegacy<PtsCanvasLegacyProps> {
  start(bound: Bound, space: CanvasSpace): void {
      window.addEventListener('keyup', e => {
        if (e.key === 'Enter') {
          // this won't work
          // this.space?.element.click(); 

          // works when passing space.pointer into the mouse event
          const p = this.space?.pointer;
          const event = new MouseEvent('click', {
            view: null,
            bubbles: true,
            clientX: p?.x,
            clientY: p?.y
          });
          this.space?.element.dispatchEvent(event);
        }
      });
  }

  animate(time: number) {
    this.form?.point( this.space?.pointer || new Pt(), 5 )
  }

  action(type: string, px: number, py: number, evt: Event): void {
    if (type==='click') console.log( px, py );
  }
}
```

Lastly, for the functional component, we can fix it in the same way --

```tsx
const handleStart: HandleStartFn = (_bound, space, form) => {
  spaceRef = space
  formRef = form

  window.addEventListener('keyup', e => {
    if (e.key === 'Enter' && canvRef.current) {
      const event = new MouseEvent('click', {
        bubbles: true,
        clientX: space.pointer.x,
        clientY: space.pointer.y
      });
      canvRef.current.dispatchEvent(event);
    }
  })
}
```
 
What do you guys think?

tibotiber commented:
Hi both. @williamngan's solution looks good, haven't had the time to try yet but should work. I'll try to dig in my current code to understand why it works without this additional work. Very curious about it now..

As for @prkirby's PR, this was the only blocker for me I believe. A nitpick question I recall I had, that's the added value of having spaceRef and formRef instead of getting the values (the current value of the references) directly in the start handler? For me, it seems similar and I'd favour less code to maintain, but I might be missing something.

williamngan commented:
Yeah, I think this looks good. For the spaceRef/formRef, I don't have a strong opinion either way, but maybe it can be useful in some edge cases?

Let me know when we're ready and I'll merge it. Thanks so much!!

prkirby commented:
Hey @tibotiber and @williamngan , I already removed the spaceRef and formRef, decided it was just such an edge case, didn't need to clutter up the code. 

So yeah I think we should be good!

tibotiber commented:
FWIW I dug into my code a few hours to understand why I don't have the jump and came up with nothing I can repro on a minimal example, so I'll go ahead and shrug ¯\_(ツ)_/¯. 

This is great work @prkirby! Thanks for the PR.

williamngan commented:
Awesome. Thank you so much @prkirby and @tibotiber ! Merging the PR now.

@prkirby do you have a twitter account? May I credit you when I tweet this?

tibotiber commented:
Fun fact: i've finally updated the Smplrspace app to use the functional component. Smooth update, thanks again for the work @prkirby :)

williamngan commented:
It's also used in various projects like https://audiobox.metademolab.com/maker (I should start collecting these examples on the site!)

Thanks again @prkirby and @tibotiber
